src/cache-utils.ts

```diff
@@ -66,35 +103,131 @@ export const getPackageManagerInfo = async (packageManager: string) => {
   }
 };
 
-export const getCacheDirectoryPath = async (
+/**
+ * glob expanding memoized because it involves potentially very deep
+ * traversing through the directories tree
+ */
+export const memoizedCacheDependencies: Record<string, string[]> = {};
+/**
+ * Expands (converts) the string input `cache-dependency-path` to list of directories that
+ * may be project roots
+ * @param cacheDependencyPath - either a single string or multiline string with possible glob patterns
+ *                              expected to be the result of `core.getInput('cache-dependency-path')`
+ * @return list of directories and possible
+ */
+const getProjectDirectoriesFromCacheDependencyPath = async (
+  cacheDependencyPath: string
+): Promise<string[]> => {
+  let cacheDependenciesPaths: string[];
+
+  // memoize unglobbed paths to avoid traversing FS
+  const memoized = memoizedCacheDependencies[cacheDependencyPath];
+  if (memoized) {
+    cacheDependenciesPaths = memoized;
+  } else {
+    cacheDependenciesPaths = (await glob
+      .create(cacheDependencyPath)
+      .then(globber => globber.glob())) || [''];
+    memoizedCacheDependencies[cacheDependencyPath] = cacheDependenciesPaths;
+  }
+
+  const existingDirectories: string[] = cacheDependenciesPaths
+    .map(cacheDependencyPath => path.dirname(cacheDependencyPath))
+    // uniq in order to do not traverse the same directories during the further processing
+    .filter(
+      (cachePath, i, result) =>
+        cachePath != null && result.indexOf(cachePath) === i
+    )
+    .filter(
+      directory =>
+        fs.existsSync(directory) && fs.lstatSync(directory).isDirectory()
+    ) as string[];
+
+  // if user explicitly pointed out some file, but it does not exist it is definitely
+  // not he wanted, thus we should throw an error not trying to workaround with unexpected
+  // result to the whole build
+  if (existingDirectories.length === 0)
+    throw Error(
+      'No existing directories found containing `cache-dependency-path`="${cacheDependencyPath}"'
+    );
+
+  return existingDirectories;
+};
+
+/**
+ * Finds the cache directories configured for the repo if cache-dependency-path is not empty
+ * @param packageManagerInfo - an object having getCacheFolderPath method specific to given PM
+ * @param cacheDependencyPath - either a single string or multiline string with possible glob patterns
+ *                              expected to be the result of `core.getInput('cache-dependency-path')`
+ * @return list of files on which the cache depends
+ */
+const getCacheDirectoriesFromCacheDependencyPath = async (
   packageManagerInfo: PackageManagerInfo,
-  packageManager: string
-) => {
-  const stdOut = await getCommandOutput(
-    packageManagerInfo.getCacheFolderCommand
+  cacheDependencyPath: string
+): Promise<string[]> => {
+  const projectDirectories = await getProjectDirectoriesFromCacheDependencyPath(
+    cacheDependencyPath
   );
+  const cacheFoldersPaths = await Promise.all(
```

Comment:
Yes, i would have done it in this way but once again - "inlining is preferable"

Please keep in mind, this PR likely to be a temporary solution because the inherited approach  "one PM per repo" is not true for "multiple directory repo" and should be re-thought from scratch. This is why i prefer to stay with the accustomed coding style.